sources/niUI/src/VideoDecoderTheoraPlayer.cpp

```diff
@@ -25,70 +25,60 @@ class cVideoDecoderTheoraPlayer : public ni::cIUnknownImpl<ni::iVideoDecoder,ni:
     Ptr<iFile> _fp = apFile;
-    Ptr<cVideoDecoderTheoraPlayer> _this = this;
-    niExec_(pLoadingExecutor,_this,_fp,withAlpha) {
-      TheoraVideoClip* pTheoraVideoClip = Theora_createVideoClip(
-          _fp,
-          withAlpha ? VIDEO_OUTPUTMODE_ALPHA : VIDEO_OUTPUTMODE,
-          2,
-          false);
-      if (!pTheoraVideoClip) {
-        niError(niFmt("Can't create the video clip '%s'.",_fp->GetSourcePath()));
-        return ni::eFalse;
-      }
+    mpTheoraVideoClip = Theora_createVideoClip(
+      _fp,
+      withAlpha ? VIDEO_OUTPUTMODE_ALPHA : VIDEO_OUTPUTMODE,
+      4,
+      false,
+      false);
+
+    if (!mpTheoraVideoClip) {
+      niError(niFmt("Can't create the video clip '%s'.",_fp->GetSourcePath()));
+      return;
+    }
 
-      niExec(Main,_this,pTheoraVideoClip,_fp,withAlpha)  {
-        _this->mpTheoraVideoClip = pTheoraVideoClip;
+    if (!mpTheoraVideoClip->load(_fp)) {
+      niError("Couldn't load the video clip.");
+      return;
+    }
 
-        if (niFlagIs(_this->mFlags,eVideoDecoderFlags_TargetTexture)) {
-          Ptr<iTexture> ptrTex = _this->mptrGraphics->CreateTexture(
-              NULL,
-              eBitmapType_2D,
-              VIDEO_PIXELFORMAT,
-              0,
-              pTheoraVideoClip->getWidth(), pTheoraVideoClip->getHeight(), 0,
-              eTextureFlags_Dynamic);
-          if (!ptrTex.IsOK()) {
-            niError(niFmt(_A("Can't create target texture for video '%s'."),_fp->GetSourcePath()));
-            return ni::eFalse;
-          }
-          _this->mpTargetTexture->mptrTexture = ptrTex;
-        }
+    if (niFlagIs(mFlags,eVideoDecoderFlags_TargetTexture)) {
+      Ptr<iTexture> ptrTex = mptrGraphics->CreateTexture(
+        NULL,
+        eBitmapType_2D,
+        VIDEO_PIXELFORMAT,
+        0,
+        mpTheoraVideoClip->getWidth(), mpTheoraVideoClip->getHeight(), 0,
+        ni::eTextureFlags_Dynamic|ni::eTextureFlags_Overlay);
+
+      if (!ptrTex.IsOK()) {
+        niError(niFmt(_A("Can't create target texture for video '%s'."),_fp->GetSourcePath()));
+        return;
+      }
+      mpTargetTexture->mptrTexture = ptrTex;
+    }
 
-        niLog(Info,niFmt(
+    niLog(Info,niFmt(
             "Video created: '%s' %s, %dx%d, %gsecs, %gfps\n",
             _fp->GetSourcePath(),
             withAlpha ? "rgba" : "rgb",
-            pTheoraVideoClip->getWidth(),
-            pTheoraVideoClip->getHeight(),
-            _this->GetLength(),
-            _this->GetVideoFps()));
-
-        _this->SetNumLoops(_this->mnNumLoops);
-        _this->SetSpeed(_this->mfSpeed);
-        _this->SetPause(_this->mbPaused);
-
-        return ni::eTrue;
-      };
-
-      return ni::eTrue;
-    };
+            mpTheoraVideoClip->getWidth(),
+            mpTheoraVideoClip->getHeight(),
+            GetLength(),
+            GetVideoFps()));
 
+    SetNumLoops(mnNumLoops);
+    SetSpeed(mfSpeed);
+    SetPause(mbPaused);
+    mpExecutor = GetConcurrent()->CreateExecutorThreadPool(1);
```

Comment:
Thats a fairly big change - probably should be passed as parameter if its necessary to use another executor in some cases. Also if the solution is for this to have its own thread youd use ThreadRun instead.

But fundamentally the issue here is that the player is designed to be low priority on purpose and to slow down if the cpu/decoding cant keep up. Thats why it uses the shared executor.

If you need it more control over the executor - like you did here - we can augment the API to take an explicit executor, but the default should remain as it was before. I think the video decoder has a global instance aswell, so that might need to be parametrized aswell although I think in this case I think SetGlobalInstance already does the job.

Reply:
I think we need a mptrExecutor->Shutdown() in the destructor when the video decoder has its own executor. There's no reason to keep a decoding thread in the bg when the decoder is dead. - The destructor of the executor might/probably already does cleanup so it could be ok as-is... would be good to double check.

Reply:
this single thread executor here is to make sure all the decoding runnable are in the same thread, before tasks will be on different threads, so we use a lot of flags to check if the last updates is finished...

Reply:
cause the decoding library is not designed to run on multiple threads

Reply:
I'm aware of that, the issue is that Overlay textures cant be tiled and dont support other various features on some hardware - especially mobile. That's the reason it's an explicit separated flag.